### PR TITLE
Use u32 for Vsock related buffer sizes

### LINKED ISSUE #4627: Use `u32` to describe vsock related buffer sizes
In #4556, we updated our generic virtio buffer handling code to describe buffer sizes using `u32` instead of `usize`, since the virtio specification states that lengths of virtio-buffers fit into `u32` (and header fields in the virtio protocol describing lengths are typed to be 32 bit integers). 

From that work, a small follow up is possible in the vsock module, where we can now also describe lengths of vsock packages using `u32` instead of `usize`. This will allow us to get rid of a few more explicit casts. Particularly, we can do the following changes to function signatures:

- `VsockConnection::peer_avail_credit` can return `u32`
- `VsockPacket::buf_size` can return `u32`
- `VsockPacket::read_at_offset_from` should be `(&self, &mut T, u32, u32) -> Result<u32, VsockError>` (e..g. the conversion to usize should only happen right around the `write_volatile_at` call
- ditto for `VsockPacket::write_from_offset_to`
Maybe we can even get away with changing the arguments `IoVecBuffer[Mut]::{read,write}_volatile_at` to `u32`s

### PR DESCRIPTION
**Pull request by RiverPhillips**
## Changes

Use u32 for all Vsock reletated buffers  instead of usize, since the Virtio specification states that lengths of virtio-buffers fit into u32

## Reason

Close #4627 

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following Developer
Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this
  PR.
- [ ] API changes follow the [Runbook for Firecracker API changes][2].
- [ ] User-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.
- [ ] New `TODO`s link to an issue.
- [ ] Commits meet
  [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

______________________________________________________________________

- [ ] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: src/vmm/src/devices/virtio/vsock/csm/connection.rs [hunk 1]
```diff
@@ -635,7 +635,7 @@ where
         };
         // Move the "forwarded bytes" counter ahead by how much we were able to send out.
         // Safe to unwrap because the maximum value is pkt.len(), which is a u32.
-        self.fwd_cnt += wrap_usize_to_u32(written);
+        self.fwd_cnt += wrap_usize_to_u32(written.try_into().unwrap());
```
Comment (roypat) [inline src/vmm/src/devices/virtio/vsock/csm/connection.rs:638]:
```suggestion
        self.fwd_cnt +=written;
```

**Review by roypat — COMMENTED** [review]
Hi, thanks for picking this up! After looking at this and #4637 again, I'm starting to think that changing everything in iovec.rs to u32 might not be that good an idea after all, with how much try_into() it introduces because of usize<->u32 conversations (the main goal for this exercise of using u32 was that I was hoping we'd eliminate try_into()s). 

Could you drop all changes outside of the `vsock` module from this PR, and do the usize<->u32 thing inside `{write,read}_from_off_to` in `packet.rs`, without any changes in iovec.rs? It should look something like 

```rs
pub fn read_at_offset_from<T: ReadVolatile + Debug>(
    &mut self,
    src: &mut T,
    offset: u32,
    count: u32,
) -> Result<u32, VsockError> {
    match self.buffer {
        VsockPacketBuffer::Tx(_) => Err(VsockError::UnwritableDescriptor),
        VsockPacketBuffer::Rx(ref mut buffer) => {
            if count
                > buffer.len()
                    .saturating_sub(VSOCK_PKT_HDR_SIZE)
                    .saturating_sub(offset)
            {
                return Err(VsockError::GuestMemoryBounds);
            }

            buffer
                .write_volatile_at(src, (offset + VSOCK_PKT_HDR_SIZE) as usize, count as usize)
                .map_err(|err| VsockError::GuestMemoryMmap(GuestMemoryError::from(err)))
                .and_then(|read| read.try_into().map_err(|_| VsockError::DescChainOverflow))
        }
    }
}
```

As an aside (I can't leave a comment on it), in connection.rs in line 241, you can do

```
diff --git a/src/vmm/src/devices/virtio/vsock/csm/connection.rs b/src/vmm/src/devices/virtio/vsock/csm/connection.rs
index 9fe744058..c1ed4eeb0 100644
--- a/src/vmm/src/devices/virtio/vsock/csm/connection.rs
+++ b/src/vmm/src/devices/virtio/vsock/csm/connection.rs
@@ -237,8 +237,7 @@ where
                         // length of the read data.
                         // Safe to unwrap because read_cnt is no more than max_len, which is bounded
                         // by self.peer_avail_credit(), a u32 internally.
-                        pkt.set_op(uapi::VSOCK_OP_RW)
-                            .set_len(u32::try_from(read_cnt).unwrap());
+                        pkt.set_op(uapi::VSOCK_OP_RW).set_len(read_cnt);
                         METRICS.rx_bytes_count.add(read_cnt as u64);
                     }
                     self.rx_cnt += Wrapping(pkt.len());
```

:)

**Comment by RiverPhillips** [comment]
Sorry, I'll try and fix this, looks like I forgot the DCO and the rebase changed a bunch more commits than mine

**Review by roypat — APPROVED** [review]
Thanks! Sorry for the late response, I was out most of last week. I've gone ahead and squashed your commits, everything else looks good to me!

**Review by kalyazin — APPROVED** [review]

**Pull request merged**